File: 138. Copy List with Random Pointer.cpp
/*
A linked list is given such that each node contains an additional random pointer which could point to any node in the list or null.

Return a deep copy of the list.
*/
#include <iostream>
#include <unordered_map>

using namespace std;

class Solution {
    public:

        struct RandomListNode {
            int label;
            RandomListNode *next, *random;
            RandomListNode(int x) : label(x), next(NULL), random(NULL) {}
        };

        RandomListNode *copyRandomList_Map(RandomListNode *head) {
            if(!head) return NULL;
            unordered_map<RandomListNode*, RandomListNode*> mp;
            RandomListNode* cur = head;
            while(cur) {
                mp[cur] = new RandomListNode(cur->label);
                cur = cur->next;
            }
            cur = head;
            while(cur) {
                mp[cur]->next = mp[cur->next];
                mp[cur]->random = mp[cur->random];
                cur = cur->next;
            }
            return mp[head];
        }

        RandomListNode *copyRandomList(RandomListNode *head) {
            if(!head) return NULL;
            RandomListNode* cur = head, *next;
            // copy node and put it after original node
            while(cur) {
                next = cur->next;
                cur->next = new RandomListNode(cur->label);
                cur->next->next = next;
                cur = next;
            }
            cur = head;
            // copy random 
            while(cur) {
                if(cur->random) {
                    cur->next->random = cur->random->next;
                }
                cur = cur->next->next;
            }

            // restore original list and format copied list
            cur = head;
            RandomListNode* fakeHead = new RandomListNode(0);
            RandomListNode* copyCur = fakeHead, *copy;
            while(cur) {
                next = cur->next->next;

                copy = cur->next;
                copyCur->next = copy;
                copyCur = copy;

                cur->next = next;
                cur = next;
            }
            return fakeHead->next;
        }
};

int main() {
    Solution sol;
    return 0;
}

/*


*/

File: py/138_Copy_List_with_Random_Pointer.py
# Definition for a Node.
class Node:
    def __init__(self, val, next, random):
        self.val = val
        self.next = next
        self.random = random
class Solution:
    def copyRandomList(self, head: 'Node') -> 'Node':
        if not head:
            return
        tmp = head
        while tmp:
            newNode = Node(tmp.val, None, None)
            newNode.next = tmp.next
            tmp.next = newNode
            tmp = tmp.next.next
        tmp = head
        while tmp:
            if tmp.random:
                tmp.next.random = tmp.random.next
            tmp = tmp.next.next
        newHead = head.next
        pold = head
        pnew = newHead
        while pnew.next:
            pold.next = pnew.next
            pold = pold.next
            pnew.next = pold.next
            pnew = pnew.next
        pold.next = None
        pnew.next = None
        return newHead

